### PR TITLE
Milestone widget: Fix CSS styles in Customizer (widget preview)

### PR DESCRIPTION
**Pull request by ivan-ottinger**
This PR fixes an issue where the styles don't load in the Customizer (widget preview in the sidebar) until the changes are saved.

  Before  |  After
:-------------------------:|:-------------------------:
Markup on 2021-11-30 at 13:22:08 | Markup on 2021-11-30 at 13:23:11

#### Changes proposed in this Pull Request:

The legacy widgets (like Milestone) are previewed in the Customizer with the help of an `iframe`. This iframe doesn't have access to styles set in a way that some of the legacy widgets use. Therefore, the widget preview is missing styles until we save the widget settings and reload Customizer (when the styles are loaded in the iframe as well).

The proposed change enqueues default CSS styles for the Milestone widget. These new styles come from a newly-created file `milestone-widget.css`. The preview iframe has access to these styles even before the widget is saved.

The new default CSS styles rely on CSS variables with fallback. This means that the Milestone widget will load the default styles until the CSS variables become available (when the user saves the changes in Customizer and reloads the page / comes back to the Customizer later).

Here is an example to better illustrate:

##### How the proposed changes work (an example with the `Button 2` theme)
1. When we add the Milestone block to a site, the default styles (defined in `milestone-widget.css`) are loaded first. These styles include a couple of default fallback colors and are used for the block preview in the left sidebar of the Customizer.

before-saved

2. As soon as we save the changes and reload the Customizer, the colors defined in the widget itself become available and are used instead.

after-saved

Please note that:
- not all themes have custom colors set
- WPCOM has an extra layer of color overrides and Customizer can also have an additional color settings set.

Examples of how the colors can be overridden on multiple levels:

  Example 1  |  Example 2  |  Example 3
:-------------------------:|:-------------------------:|:-------------------------:
fix-in-action | Markup on 2021-12-02 at 09:55:58 | Markup on 2021-12-02 at 10:02:38

As a result, the behavior of each theme may be slightly different (plus there's a difference between WPCOM site and a Jetpack site that is running the same theme). But in all cases, at least the default styles should be loaded (as can be seen in the "After" screenshot above).

The proposed changes also fix the view on the widgets page (`wp-admin/widgets.php`).

---

#### Jetpack product discussion
This PR is part of the Legacy Widgets Migration/Fix project (pdf5j4-4C-p2).

#### Does this pull request change what data or activity we track or use?
No, it does not.

#### Testing instructions:

General steps:
1. Pick a theme you like (you can test multiple)
2. Navigate to Customizer → Widgets
3. Add the Milestone widget
4. Adjust the widget settings if you like
5. Click away from the widget without saving the changes
6. The widget preview should load properly with default colors
7. Save the changes and reload the page; custom theme colors (if available) should be applied

Before the Milestone widget is added to the site, there should be no Milestone-related CSS enqueued. We can check for this in the browser inspector searching for `milestone-widget.css` or `milestone-widget-css` references. Only after the widget is added to the site, the styles should load in a form similar to the following:

```html
<link rel="stylesheet" id="milestone-widget-css" href="http://localhost/wp-content/plugins/jetpack/modules/widgets/milestone/milestone-widget.css?ver=10.5-a.0" media="all">
```

or

```html
<link rel="stylesheet" id="milestone-widget-css" href="https://randomsite.random/wp-content/mu-plugins/widgets/milestone/milestone-widget.css?m=1638543000h&amp;ver=wpcom" media="all">
```

When testing on the WPCOM sandbox, it is necessary to sandbox the following:
- the site itself
- `public-api.wordpress.com`
- `s0.wp.com`
- `s1.wp.com`
- `s2.wp.com`

### CODE REVIEW

File: projects/plugins/jetpack/modules/widgets/milestone/class-milestone-widget.php [hunk 1]
```diff
@@ -145,10 +149,10 @@ public static function styles_template() {
...
 .milestone-countdown,
 .milestone-message {
-	background-color: <?php echo self::sanitize_color_hex( $colors['bg'] ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>;
-	border: 1px solid <?php echo self::sanitize_color_hex( $colors['border'] ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>;
+	background-color: var(--milestone-bg-color);
+	border: 1px solid var(--milestone-border-color);
 	border-top: 0;
-	color: <?php echo self::sanitize_color_hex( $colors['text'] ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>;
+	color: var(--milestone-text-color);
 	padding-bottom: 1em;
```
Comment (yansern) [inline projects/plugins/jetpack/modules/widgets/milestone/class-milestone-widget.php:156]: Should these also be moved to the CSS file? (also, I did not highlight those were above). Was thinking that the only thing that the PHP end was gonna generate was the CSS var declarations.

**Review by yansern — COMMENTED** [review]
Should it not be where the declaration of the CSS vars are the only generated part by the PHP end and the rest goes into the CSS file?.

https://github.com/Automattic/jetpack/pull/21904/files#diff-3aae433c761c73c82488e13a711282906752962e15a67e398d79fd8c9525be91R116-R161

**Comment by ivan-ottinger** [comment]
> Should it not be where the declaration of the CSS vars are the only generated part by the PHP end and the rest goes into the CSS file?.

Thanks, Yan! Good point. I have removed the duplicate styles from `class-milestone-widget.php` in https://github.com/Automattic/jetpack/pull/21904/commits/5c4acd9d14642b98613b581ec217a76523e2aca9.

File: projects/plugins/jetpack/modules/widgets/milestone/class-milestone-widget.php [hunk 2]
```diff
@@ -115,44 +116,9 @@ public static function styles_template() {
 <style>
 .milestone-widget {
 	margin-bottom: 1em;
```
Comment (yansern) [inline projects/plugins/jetpack/modules/widgets/milestone/class-milestone-widget.php:118]:
```suggestion
```

One more line to remove and I think we're good. :)

**Review by yansern — CHANGES_REQUESTED** [review]

File: projects/plugins/jetpack/modules/widgets/milestone/class-milestone-widget.php (hunk 2 above)
Comment (ivan-ottinger) [reply projects/plugins/jetpack/modules/widgets/milestone/class-milestone-widget.php]: oops! Missed that one. Fixed in: https://github.com/Automattic/jetpack/pull/21904/commits/dacc296cb993b37911e325b005eb87786744cc47. Cheers! :)

**Review by yansern — DISMISSED** [review]
I think this is good to go considering the circumstances we're in (theme colors may not exist, wpcom has overrides, color might not show til saved). Given that there's fallback default color within the CSS file itself, e.g.` var(--milestone-text-color, #111);`, I think enqueuing the CSS file is a good fix. Good to ship.

File: projects/plugins/jetpack/modules/widgets/milestone/class-milestone-widget.php [hunk 3]
```diff
@@ -49,6 +49,7 @@ public function __construct() {
 		);
 
 		add_action( 'wp_enqueue_scripts', array( __class__, 'enqueue_template' ) );
+		add_action( 'wp_enqueue_scripts', array( $this, 'enqueue_scripts' ) );
```
Comment (jeherve) [inline projects/plugins/jetpack/modules/widgets/milestone/class-milestone-widget.php:52]: I would recommend against doing this, as this will mean enqueuing the stylesheet on all pages of a site as soon as the Widgets module is active, even if they don't use that specific widget at all.

One thing you can do instead is only enqueuing the stylesheet if the widget is active, like it's done for this widget for example:
https://github.com/Automattic/jetpack/blob/e13487d0dbbe4c743739b6f1a3e37c143a8c2614/projects/plugins/jetpack/modules/widgets/authors.php#L32-L34

File: projects/plugins/jetpack/modules/widgets/milestone/class-milestone-widget.php [hunk 4]
```diff
@@ -283,6 +248,18 @@ public function widget( $args, $instance ) {
 		do_action( 'jetpack_stats_extra', 'widget_view', 'milestone' );
 	}
 
+	/**
+	 * Enqueue widget styles
+	 */
+	public function enqueue_scripts() {
+		wp_enqueue_style(
+			'milestone-widget',
```
Comment (jeherve) [inline projects/plugins/jetpack/modules/widgets/milestone/class-milestone-widget.php:256]: Since you're creating a new stylesheet, you could opt to include it in the concatenated list of frontend stylesheets included in Jetpack, here:
https://github.com/Automattic/jetpack/blob/96351c0f811f5e1cb658aad1fbc0c441a906af04/projects/plugins/jetpack/class.jetpack.php#L55-L60
https://github.com/Automattic/jetpack/blob/e6f3dc5171e4f23b5e835746b25733ab71715aa8/projects/plugins/jetpack/tools/builder/frontend-css.js#L20-L25

In practice, this means that on most sites, the stylesheet you've created would never be enqueued; instead the css would get included as part of a general `jetpack.css` stylesheet.

**Review by jeherve — CHANGES_REQUESTED** [review]

File: projects/plugins/jetpack/modules/widgets/milestone/class-milestone-widget.php (hunk 3 above)
Comment (arcangelini) [reply projects/plugins/jetpack/modules/widgets/milestone/class-milestone-widget.php:52]: While working on the `eu_cookie_law_widget` I had success putting the enqueue in a function and calling it in the `widget()` function. Not sure if the same would apply here though 🤷‍♂️

Comment (ivan-ottinger) [reply projects/plugins/jetpack/modules/widgets/milestone/class-milestone-widget.php:52]: Thank you for your feedback Jeremy & Tony.

When I try to enqueue the styles conditionally in the constructor: 

```php
if ( is_active_widget( false, false, $this->id_base, true ) || is_active_widget( false, false, 'monster', true ) || is_customize_preview() ) {
	add_action( 'wp_enqueue_scripts', array( $this, 'enqueue_scripts' ) );
}
```
or when I try to load them (with the same condition) in the `widget()` function, these styles are not loaded in the Legacy Widget block `iframe` unfortunately.

Markup on 2021-12-03 at 15:03:32

It is not clear to me how this could be resolved yet. 🤔

File: projects/plugins/jetpack/modules/widgets/milestone/class-milestone-widget.php (hunk 4 above)
Comment (ivan-ottinger) [reply projects/plugins/jetpack/modules/widgets/milestone/class-milestone-widget.php:256]: That's good to know! Thanks! I have applied the changes in the commit https://github.com/Automattic/jetpack/pull/21904/commits/79a650d4cc15bb0bd2cb40e1510924dc910b5b9e.

File: projects/plugins/jetpack/modules/widgets/milestone/class-milestone-widget.php (hunk 3 above)
Comment (jeherve) [reply projects/plugins/jetpack/modules/widgets/milestone/class-milestone-widget.php:52]: >  these styles are not loaded in the Legacy Widget block iframe unfortunately.

This may be because `is_customize_preview()` doesn't cover all our possible use-cases here?

Would something like this work better?

https://github.com/Automattic/jetpack/blob/81ea36abcd016aefb9d062f047244340b5a1a100/projects/plugins/jetpack/modules/widgets/contact-info.php#L43-L48

Comment (ivan-ottinger) [reply projects/plugins/jetpack/modules/widgets/milestone/class-milestone-widget.php:52]: It doesn't work either unfortunately.  :/

I tried a couple of things and looked at [how the iframe preview is built in core/Gutenberg](https://github.com/WordPress/gutenberg/blob/d7d2484611fc15e407e1cf2b786b76799320c8b4/packages/widgets/src/blocks/legacy-widget/index.php#L98-L152), but no progress so far.

I will get back to this next week. Maybe it would be worth to build a new Milestone block, hehe. :)

Comment (ivan-ottinger) [reply projects/plugins/jetpack/modules/widgets/milestone/class-milestone-widget.php]: Alright. I did a couple of tests and:

When the CSS script is enqueued in the `widget()` - as can be seen in the latest commit (https://github.com/Automattic/jetpack/pull/21904/commits/19060f42f07a3550ce1a6c0aeb37a93353919c28), the CSS is enqueued only when the widget is added to the site. This is true even when we don't enclose the enqueue call in a condition.

So circling back to the original @arcangelini's suggestion above. :)

Similar usage can be found in the `gravatar-profile` widget:

https://github.com/Automattic/jetpack/blob/135bb0b766e4410b1893feedc7f14ddbe94a86a7/projects/plugins/jetpack/modules/widgets/gravatar-profile.php#L97

---

In other words, everything seems to be working as expected now.

**Comment by arcangelini** [comment]
Maybe I am missing something, but when I tested this the styling is not matching the preview? It looks great in the Customizer widget preview but not on the front-end or the customizer site preview. These were my results from testing on my Jetpack test environment.

Markup 2021-12-07 at 20 44 14

**Comment by ivan-ottinger** [comment]
> Maybe I am missing something, but when I tested this the styling is not matching the preview? It looks great in the Customizer widget preview but not on the front-end or the customizer site preview. These were my results from testing on my Jetpack test environment.

Thank you for testing Tony! That is the `Twenty Twenty-One` theme, right?

It is working correctly when I test it on my local Jetpack environment. 🤔 Here are the steps I take (in case we are doing things differently):
https://user-images.githubusercontent.com/25105483/145176110-821bc81e-b220-4f2f-b0b2-76fcd90aac8c.mp4

(I also tried to follow the steps with freshly-cleared browser cache and three browsers (Chrome, Ghost and Firefox) - just in case. It worked the same in all three cases)

**Review by jeherve — APPROVED** [review]
This tests well for me. 👍

File: projects/plugins/jetpack/tools/builder/frontend-css.js [hunk 5]
```diff
@@ -52,6 +52,7 @@ export const frontendCSSSeparateFilesList = [
 	'modules/widgets/search/css/search-widget-frontend.css',
 	'modules/widgets/simple-payments/style.css',
 	'modules/widgets/social-icons/social-icons.css',
+	'modules/widgets/milestone/milestone-widget.css',
```
Comment (yansern) [inline projects/plugins/jetpack/tools/builder/frontend-css.js:55]: 👍

**Review by jeherve — APPROVED** [review]

**Review by yansern — APPROVED** [review]

File: projects/plugins/jetpack/modules/widgets/milestone/class-milestone-widget.php (hunk 3 above)
Comment (yansern) [reply projects/plugins/jetpack/modules/widgets/milestone/class-milestone-widget.php]: Just want to say thanks all for chiming in and sharing knowledge on enqueueing stylesheets. :)

**Comment by ivan-ottinger** [comment]
Thanks so much folks! I will commit and deploy to WPCOM in a couple of moments.

**Comment by ivan-ottinger** [comment]
Successfully deployed to WPCOM. Changeset: r236308-wpcom.
